### ISSUE TITLE
My elastic search bulk indexer is not properly closed

### ISSUE THREAD

**Issue by girafferiel** [elastic/go-elasticsearch#285]
I'm using AWS ElasticSearch 7.10 with Kafka and pushed some code to my development environment. The configuration I used was:

```
func NewClient(addresses []string, username string, password string) (*elasticsearch.Client, error) {
    es, err := elasticsearch.NewClient(elasticsearch.Config{
        Addresses:     addresses,
        Username:      username,
        Password:      password,
        RetryOnStatus: []int{502, 503, 504, 429},
        MaxRetries:    5,
    })
    if err != nil {
        return nil, err
    }

    _, err = es.Info()
    if err != nil {
        return nil, err
    }
    return es, nil
}

func NewBulkIndexer(es *elasticsearch.Client, defaultIndexName string) (esutil.BulkIndexer, error) {
    bi, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
        Index:         defaultIndexName, // The default index name
        Client:        es,               // The Elasticsearch client
        NumWorkers:    runtime.NumCPU(), // The number of worker goroutines
        FlushBytes:    int(5e+6),        // The flush threshold in bytes
        FlushInterval: 30 * time.Second, // The periodic flush interval
    })
    return bi, err
}
```

```
// repository/job.go
func (r *JobRepo) SyncJob(ctx context.Context, indexName string, job domain.Job) error {
    body := new(bytes.Buffer)
    if err := json.NewEncoder(body).Encode(job); err != nil {
        logger.Errorf("Error parsing task: %v", err)
        return err
    }

    err := r.bi.Add(
        context.Background(),
        esutil.BulkIndexerItem{
            Index:      indexName,
            Action:     "index",
            DocumentID: job.ID,
            Body:       bytes.NewReader(body.Bytes()),
            OnSuccess: func(ctx context.Context, item esutil.BulkIndexerItem, res esutil.BulkIndexerResponseItem) {
                logger.Infof("Success syncing job ID: %s", job.ID)
            },
            OnFailure: func(ctx context.Context, item esutil.BulkIndexerItem, res esutil.BulkIndexerResponseItem, err error) {
                if err != nil {
                    logger.Errorf("Bulk indexer error: %v", err)
                } else {
                    logger.Errorf("Bulk indexer error: %s: %s", res.Error.Type, res.Error.Reason)
                }
            },
        },
    )
    if err != nil {
        logger.Errorf("Unexpected error: %s", err)
        return err
    }

    return nil
}
```
Only after some period I added the code to close my bulk indexer:

```
func (r *JobRepo) Disconnect() error {
if err := r.bi.Close(context.Background()); err != nil { logger.Errorf("Unexpected error: %v", err) return err } return nil }

// main.go

    addresses := strings.Split(cfg.Elastic.Addresses, ",")

    es, err := elasticsearch.NewClient(addresses, cfg.Elastic.Username, cfg.Elastic.Password)
    if err != nil {
        logger.Errorf("Failed to create Elasticsearch client: %v", err)
        return err
    }

    bi, err := elasticsearch.NewBulkIndexer(es, cfg.IndexName)
    if err != nil {
        logger.Errorf("Failed to create bulk indexer %v", err)
        return err
    }

    repo := repository.NewJobRepo(es, bi, cfg.IndexName)

    defer cleanup(repo)

    c := make(chan os.Signal, 1)
    signal.Notify(c, syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL)
    go func() {
        logger.Infof("Exiting due to signal: %v", <-c)
        cleanup(repo)
        os.Exit(1)
    }()
.
.
.
.
.
func cleanup(repo *repository.JobRepo) error {
    return repo.Disconnect()
}
```
Although I have pushed environment, after every launch there are logs like these:

Success syncing job ID: ....

more than 50 entries. I thought I have closed the bulk indexer properly but it still show these messages although I haven't publish any kafka message to sync my records. Is there anything I can do? Im thinking of deleting my aws elastic db since the contents arent important and I want to start over using new index template anyway but I want to know other way to fix this bulk indexer. thanks

**Comment by ljluestc** [comment]
```
package main

import (
    "bytes"
    "context"
    "encoding/json"
    "log"
    "os"
    "os/signal"
    "runtime"
    "strings"
    "syscall"
    "time"

    "github.com/elastic/go-elasticsearch/v7"
    "github.com/elastic/go-elasticsearch/v7/esutil"
)

// Config holds configuration for Elasticsearch
type Config struct {
    Elastic struct {
        Addresses string
        Username  string
        Password  string
        IndexName string
    }
}

// Job represents the structure of the job data
type Job struct {
    ID string `json:"id"`
    // Add other fields as needed
}

// JobRepo holds the Elasticsearch client and bulk indexer
type JobRepo struct {
    Client *elasticsearch.Client
    BulkIndexer esutil.BulkIndexer
    IndexName string
}

// NewClient initializes and returns a new Elasticsearch client
func NewClient(addresses []string, username, password string) (*elasticsearch.Client, error) {
    es, err := elasticsearch.NewClient(elasticsearch.Config{
        Addresses:     addresses,
        Username:      username,
        Password:      password,
        RetryOnStatus: []int{502, 503, 504, 429},
        MaxRetries:    5,
    })
    if err != nil {
        return nil, err
    }

    _, err = es.Info()
    if err != nil {
        return nil, err
    }
    return es, nil
}

// NewBulkIndexer initializes and returns a new BulkIndexer
func NewBulkIndexer(es *elasticsearch.Client, defaultIndexName string) (esutil.BulkIndexer, error) {
    bi, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
        Index:         defaultIndexName,
        Client:        es,
        NumWorkers:    runtime.NumCPU(),
        FlushBytes:    int(5e+6),
        FlushInterval: 30 * time.Second,
    })
    return bi, err
}

// NewJobRepo creates and returns a new JobRepo
func NewJobRepo(es *elasticsearch.Client, bi esutil.BulkIndexer, indexName string) *JobRepo {
    return &JobRepo{
        Client:     es,
        BulkIndexer: bi,
        IndexName:  indexName,
    }
}

// SyncJob syncs a single job to Elasticsearch using BulkIndexer
func (r *JobRepo) SyncJob(ctx context.Context, job Job) error {
    body := new(bytes.Buffer)
    if err := json.NewEncoder(body).Encode(job); err != nil {
        log.Printf("Error encoding job: %v", err)
        return err
    }

    err := r.BulkIndexer.Add(
        ctx,
        esutil.BulkIndexerItem{
            Index:      r.IndexName,
            Action:     "index",
            DocumentID: job.ID,
            Body:       bytes.NewReader(body.Bytes()),
            OnSuccess: func(ctx context.Context, item esutil.BulkIndexerItem, res esutil.BulkIndexerResponseItem) {
                log.Printf("Success syncing job ID: %s", job.ID)
            },
            OnFailure: func(ctx context.Context, item esutil.BulkIndexerItem, res esutil.BulkIndexerResponseItem, err error) {
                if err != nil {
                    log.Printf("Bulk indexer error: %v", err)
                } else {
                    log.Printf("Bulk indexer error: %s: %s", res.Error.Type, res.Error.Reason)
                }
            },
        },
    )
    if err != nil {
        log.Printf("Unexpected error: %s", err)
        return err
    }
    return nil
}

// Disconnect closes the bulk indexer
func (r *JobRepo) Disconnect() error {
    if r.BulkIndexer != nil {
        if err := r.BulkIndexer.Close(context.Background()); err != nil {
            log.Printf("Error closing BulkIndexer: %v", err)
            return err
        }
    }
    return nil
}

func main() {
    var cfg Config
    // Assume cfg is populated from config file or environment variables

    addresses := strings.Split(cfg.Elastic.Addresses, ",")
    es, err := NewClient(addresses, cfg.Elastic.Username, cfg.Elastic.Password)
    if err != nil {
        log.Fatalf("Failed to create Elasticsearch client: %v", err)
    }

    bi, err := NewBulkIndexer(es, cfg.Elastic.IndexName)
    if err != nil {
        log.Fatalf("Failed to create bulk indexer: %v", err)
    }

    repo := NewJobRepo(es, bi, cfg.Elastic.IndexName)

    // Setup signal handler for graceful shutdown
    sigChan := make(chan os.Signal, 1)
    signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL)

    go func() {
        sig := <-sigChan
        log.Printf("Received signal: %v", sig)
        cleanup(repo)
        os.Exit(1)
    }()

    // Simulate some work or integration with Kafka
    // ...

    // Close BulkIndexer when done or on shutdown
    cleanup(repo)
}

func cleanup(repo *JobRepo) {
    if err := repo.Disconnect(); err != nil {
        log.Printf("Error during cleanup: %v", err)
    }
}
```

**Comment by MattDevy** [comment]
Hi @girafferiel , sorry for the long wait? Did you get sorted here?

From the information I can see, it looks like maybe you're not acknowledging some kafka messages somewhere and they are being redelivered when you restart your application? 

Let me know if there's any more info you can provide here. Thank you!

**Comment by MattDevy** [comment]
Closing this out due to inactivity. If you're still experiencing this on a current version, feel free to open a new issue with details.
